edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -132,63 +132,153 @@ public async Task SetEndpoint(Endpoint newEndpoint, IList<uint> priorities)
+                        foreach (DictionaryEntry entry in snapshot)
+                        {
+                            // Also check for cancellation in every inner loop,
+                            // since it could take time to send a batch of messages
+                            if (this.cts.IsCancellationRequested)
+                            {
+                                break;
+                            }
+
+                            uint priority = (uint)entry.Key;
+                            EndpointExecutorFsm fsm = (EndpointExecutorFsm)entry.Value;
+                            this.lastUsedFsm = fsm;
+
+                            (IMessageIterator, StoreMessagesProvider) pair;
+                            if (!messageProviderPairs.TryGetValue(priority, out pair))
+                            {
+                                // Create and cache a new pair for the message provider
+                                // so we can reuse it every loop
+                                pair.Item1 = this.messageStore.GetMessageIterator(GetMessageQueueId(this.Endpoint.Id, priority), fsm.Checkpointer.Offset + 1);
+                                pair.Item2 = new StoreMessagesProvider(pair.Item1, batchSize);
+                                messageProviderPairs.Add(priority, pair);
+                            }
+
+                            StoreMessagesProvider storeMessagesProvider = pair.Item2;
+                            IMessage[] messages = await storeMessagesProvider.GetMessages();
+                            if (messages.Length > 0)
+                            {
+                                await this.ProcessMessages(messages, fsm);
+                                Events.SendMessagesSuccess(this, messages);
+                                MetricsV0.DrainedCountIncrement(this.Endpoint.Id, messages.Length);
+
+                                // Only move on to the next priority if the queue for the current
+                                // priority is empty. If we processed any messages, break out of
+                                // the inner loop to restart at the beginning of the priorities list
+                                // again. This is so we can catch and process any higher priority
+                                // messages that came in while we were sending the current batch
+                                haveMessagesRemaining = true;
+                                break;
+                            }
+                        }
+
+                        if (!haveMessagesRemaining)
+                        {
+                            // All the message queues have been drained, reset the hasMessagesInQueue flag.
+                            this.hasMessagesInQueue.Reset();
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        Events.SendMessagesError(this, ex);

```

Comment:
For failed messages, it will pick up and re-try every cycle. right?

Reply:
yes, that's handled by the FSM in ProcessMessages()

---
In reply to: [391350993](https://github.com/Azure/iotedge/pull/2670#discussion_r391350993) [](ancestors = 391350993)

Reply:
follow up question: what happens if there is some reason causing a specific message throws exception during processing every time.  Will this message block all queues? This may already exist.  we can address this in another PR.

Reply:
the retry loop is not here, it's in the FSM.  This loop is just to get messages out of storage.

If the FSM encounters a message throws an exception, it's checked against a whitelist of exceptions, then the retry action is resolved against the specified RetryStrategy.  Unrecoverable messages are generally dropped on the floor.

All this is already handled by the lower layers, and the logical layer here doesn't need to deal with it.